driver/src/driver/scheduler/evm.rs

```diff
@@ -0,0 +1,279 @@
+    ///
+    /// This is the current batch ID minus 1, as the current batch ID is the ID
+    /// of the batch that is currently accepting orders.
+    fn current_solving_batch(&self) -> Result<u32> {
+        Ok(self.exchange.get_current_auction_index()? - 1)
+    }
+}
+
+impl Scheduler for EvmScheduler<'_> {
+    fn start(&mut self) -> ! {
+        loop {
+            if let Err(err) = self.step() {
+                error!("EVM scheduler error: {}", err);
+            }
+            thread::sleep(POLL_TIMEOUT);
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::contracts::stablex_contract::MockStableXContract;
+    use crate::driver::stablex_driver::MockStableXDriver;
+    use anyhow::anyhow;
+    use ethcontract::U256;
+    use mockall::predicate::eq;
+    use mockall::Sequence;
+
+    #[test]
+    fn scheduler_step() {
+        let mut exchange = MockStableXContract::new();
+        exchange
+            .expect_get_current_auction_index()
+            .returning(|| Ok(42));
+        exchange
+            .expect_get_current_auction_remaining_time()
+            .returning(|| Ok(Duration::from_secs(270)));
+
+        let mut driver = MockStableXDriver::new();
+        driver
+            .expect_run()
+            .with(eq(U256::from(41)), eq(Duration::from_secs(150)))
+            .returning(|_, _| DriverResult::Ok);
+
+        let mut scheduler = EvmScheduler::with_defaults(&exchange, &driver);
+
+        scheduler.step().unwrap();
+        assert_eq!(scheduler.last_batch, Some(41));
+    }
+
+    #[test]
+    fn scheduler_run_next_batch() {
+        let mut exchange = MockStableXContract::new();
+        exchange
+            .expect_get_current_auction_index()
+            .returning(|| Ok(42));
+        exchange
+            .expect_get_current_auction_remaining_time()
+            .returning(|| Ok(Duration::from_secs(270)));
+
+        let mut driver = MockStableXDriver::new();
+        driver.expect_run().returning(|_, _| DriverResult::Ok);
+
+        let mut scheduler = EvmScheduler::with_defaults(&exchange, &driver);
+        scheduler.last_batch = Some(40);
+
+        scheduler.step().unwrap();
+        assert_eq!(scheduler.last_batch, Some(41));
+    }
+
+    #[test]
+    fn scheduler_skips_previously_processed_batches() {
+        let mut exchange = MockStableXContract::new();
+        exchange
+            .expect_get_current_auction_index()
+            .returning(|| Ok(42));
+
+        let driver = MockStableXDriver::new();
+
+        let mut scheduler = EvmScheduler::with_defaults(&exchange, &driver);
+        scheduler.last_batch = Some(41);
+
+        scheduler.step().unwrap();
+    }
+
+    #[test]
+    fn scheduler_retries_when_batch_changes_during_run() {
+        let mut seq = Sequence::new();
+        let mut exchange = MockStableXContract::new();
+        exchange
+            .expect_get_current_auction_index()
+            .times(1)
+            .in_sequence(&mut seq)
```

Comment:
This is 🆒

Reply:
Thought you might also like: https://github.com/gnosis/dex-services/blob/ad1fc0ed780b06a6b63bae81951b348521e7e1ea/driver/src/driver/scheduler.rs#L21

You mentioned that you did not like infinite loops too much, this `!` type can be used to indicate to the compiler that this method never returns. It even gives you cool things like compiler warnings in case you write statements after calling it.